src/cargo/core/compiler/custom_build.rs

```diff
@@ -17,6 +17,22 @@ use std::sync::{Arc, Mutex};
 
 const CARGO_WARNING: &str = "cargo:warning=";
 
+macro_rules! check_and_add_target {
```

Comment:
Could this perhaps be a function instead of a macro?

(it'd be great to avoid 9-argument things with closures that close over outer values too)

Reply:
I chose macro over function as it might be a tedious to pass all the required parameters to it. I understand your concern though. If you think it's better as function I'll change it.

Reply:
I wonder what you'd expect the function to do and what would be it's signature? We can't pass the check function (as it's bound to a single instance?)

Reply:
Without being the one writing this I don't know precisely what it would look like. I find myself very rarely needing 9-parameter things and often in need of a refactoring of such a need arises.

Reply:
I think it might be better then to refactor the macro to presume it's running from the specific context.

Reply:
I tried playing with it a bit. Couldn't think of any better way to avoid the code duplication.
If you think it's better to revert to the old style (a bit of repetition) then I'm okay with it.